File: 0124.二叉树中的最大路径和/0124-二叉树中的最大路径和.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /** Maintain an int for the class, update during recursion,
    use min_value in case there is a single node with negative val. */ 
    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        gainMaxSum(root);
        return maxSum;
    }

    private int gainMaxSum(TreeNode node) {
        if (node == null) {
            return 0;
        }
        //Updating the class variable maxSum
        int nodeVal = node.val;
        int leftMax = Math.max(gainMaxSum(node.left), 0);
        int rightMax = Math.max(gainMaxSum(node.right), 0);
        int nodeMaxPath = nodeVal + leftMax + rightMax;
        maxSum = Math.max(maxSum, nodeMaxPath);
        //return the maxvalue that the node can provide for other nodes
        return node.val + Math.max(leftMax, rightMax);
    }
}

File: 0124.二叉树中的最大路径和/0124-二叉树中的最大路径和.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int maxSum = INT_MIN;

int gainMaxSum(struct TreeNode* node) {
    if (node == NULL) {
        return 0;
    }
    //Updating the class variable maxSum
    int nodeVal = node->val;
    int leftMax = fmax(gainMaxSum(node->left), 0);
    int rightMax = fmax(gainMaxSum(node->right), 0);
    int nodeMaxPath = nodeVal + leftMax + rightMax;
    maxSum = fmax(maxSum, nodeMaxPath);
    //return the maxvalue that the node can provide for other nodes
    return node->val + fmax(leftMax, rightMax);
    }

int maxPathSum(struct TreeNode* root){
    gainMaxSum(root);
    int res = maxSum;
    maxSum = INT_MIN;
    return res;
}